atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/cache/TransactionScopedCacheImplTest.java

```diff
@@ -138,6 +158,146 @@ public void allValuesReadFromCachePreventsReadToDb() {
+
+        NavigableMap<byte[], RowResult<byte[]>> lookup = getRowsCall(cache, rowsAndCols, columns);
+        assertDirectLookups(createCell(1, 3), createCell(2, 2), createCell(2, 3));
+        assertRowLookup(3);
+
+        assertExpectedResult(lookup, 3);
+
+        verify(metrics).increaseGetRowsHits(3);
+        verify(metrics).increaseGetRowsCellLookups(3);
+        verify(metrics).increaseGetRowsRowLookups(1);
+
+        clearInvocations(directLookups, rowLookups);
+        getRowsCall(cache, rowsAndCols, columns);
+        verifyNoInteractions(directLookups, rowLookups);
+    }
+
+    @Test
+    public void getRowsSkipsFullyPresentRows() {
+        // row 2 will have a row lookup
+        TransactionScopedCache cache = createCache(false, createCell(1, 1), createCell(1, 2), createCell(1, 3));
+
+        ImmutableList<byte[]> rowsAndCols = ImmutableList.of(createBytes(1), createBytes(2));
+        ColumnSelection columns = ColumnSelection.create(rowsAndCols);
+
+        NavigableMap<byte[], RowResult<byte[]>> lookup = getRowsCall(cache, rowsAndCols, columns);
+
+        assertDirectLookups();
+        assertRowLookup(2);
+        assertExpectedResult(lookup, 2);
+    }
+
+    @Test
+    public void getRowsRespectsEmptyLookups() {
+        // row 2 will have a cell lookup
+        TransactionScopedCache cache = createCache(
+                true, createCell(1, 1), createCell(1, 2), createCell(1, 3), createCell(2, 1), createCell(2, 3));
+
+        ImmutableList<byte[]> rowsAndCols = ImmutableList.of(createBytes(1), createBytes(2), createBytes(3));
+        ColumnSelection columns = ColumnSelection.create(rowsAndCols);
+
+        // rowLoader returns empty
+        NavigableMap<byte[], RowResult<byte[]>> lookup = getRowsWithEmptyRowLookups(cache, rowsAndCols, columns);
+
+        Cell onlyCell = createCell(2, 2);
+        assertDirectLookups(onlyCell);
+        assertRowLookup(3);
+        assertThat(lookup.size()).isEqualTo(1);
+        assertThat(lookup.get(createBytes(2)).getOnlyColumnValue()).containsExactly(createBytes(onlyCell));
+    }
+
+    @Test
+    public void getRowsDoesNotBreakOnEmptyRowsAndCachesThemCorrectly() {
+        TransactionScopedCache cache = TransactionScopedCacheImpl.create(snapshotWithSingleValue(), metrics);
+
+        ImmutableList<byte[]> rowsAndCols = ImmutableList.of(createBytes(2), createBytes(3));
+        ColumnSelection columns = ColumnSelection.create(rowsAndCols);
+
+        NavigableMap<byte[], RowResult<byte[]>> lookup = getRowsWithEmptyRowLookups(cache, rowsAndCols, columns);
+        verify(rowLookups, times(2)).add(any());
+
+        clearInvocations(rowLookups);
+        getRowsWithEmptyRowLookups(cache, rowsAndCols, columns);
+        assertDirectLookups();
+        verifyNoInteractions(rowLookups);
+        assertThat(lookup).isEmpty();
+    }
+
+    private TransactionScopedCache createCache(boolean emptyCells, Cell... cells) {
```

Comment:
private methods should go below all tests.

Reply:
I agree, probably missed this in the last refactor